File: backend/routes/todos.js
const express = require('express');
const router = express.Router();
const { body, param, query, validationResult } = require('express-validator');
const Todo = require('../models/Todo');

const handleValidationErrors = (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({
      success: false,
      error: 'Validation failed',
      details: errors.array()
    });
  }
  next();
};

router.get(
  '/',
  [
    query('completed').optional().isBoolean().toBoolean(),
    query('priority').optional().isIn(['low', 'medium', 'high']),
    query('sortBy').optional().isString(),
    query('order').optional().isIn(['asc', 'desc'])
  ],
  handleValidationErrors,
  async (req, res) => {
    try {
      const { completed, priority, sortBy = 'createdAt', order = 'desc' } = req.query;
      
      const filter = {};
      if (completed !== undefined) {
        filter.completed = completed;
      }
      if (priority) {
        filter.priority = priority;
      }

      const sortOrder = order === 'desc' ? -1 : 1;
      const sortOptions = { [sortBy]: sortOrder };

      const todos = await Todo.find(filter).sort(sortOptions);

      res.json({
        success: true,
        count: todos.length,
        data: todos
      });
    } catch (error) {
      console.error('Error fetching todos:', error);
      res.status(500).json({
        success: false,
        error: 'Server error while fetching todos',
        message: error.message
      });
    }
  }
);

router.get(
  '/:id',
  [
    param('id').isMongoId().withMessage('Invalid todo ID')
  ],
  handleValidationErrors,
  async (req, res) => {
    try {
      const todo = await Todo.findById(req.params.id);

      if (!todo) {
        return res.status(404).json({
          success: false,
          error: 'Todo not found'
        });
      }

      res.json({
        success: true,
        data: todo
      });
    } catch (error) {
      console.error('Error fetching todo:', error);
      res.status(500).json({
        success: false,
        error: 'Server error while fetching todo',
        message: error.message
      });
    }
  }
);

router.post(
  '/',
  [
    body('title')
      .trim()
      .notEmpty().withMessage('Title is required')
      .isLength({ min: 1, max: 200 }).withMessage('Title must be between 1 and 200 characters'),
    body('description')
      .optional()
      .trim()
      .isLength({ max: 500 }).withMessage('Description cannot exceed 500 characters'),
    body('priority')
      .optional()
      .isIn(['low', 'medium', 'high']).withMessage('Priority must be low, medium, or high')
  ],
  handleValidationErrors,
  async (req, res) => {
    try {
      const { title, description, priority } = req.body;

      const todo = new Todo({
        title,
        description: description || '',
        completed: false,
        priority: priority || 'medium'
      });

      await todo.save();

      res.status(201).json({
        success: true,
        data: todo
      });
    } catch (error) {
      console.error('Error creating todo:', error);
      res.status(400).json({
        success: false,
        error: 'Error creating todo',
        message: error.message
      });
    }
  }
);

router.put(
  '/:id',
  [
    param('id').isMongoId().withMessage('Invalid todo ID'),
    body('title')
      .optional()
      .trim()
      .notEmpty().withMessage('Title cannot be empty')
      .isLength({ min: 1, max: 200 }).withMessage('Title must be between 1 and 200 characters'),
    body('description')
      .optional()
      .trim()
      .isLength({ max: 500 }).withMessage('Description cannot exceed 500 characters'),
    body('completed')
      .optional()
      .isBoolean().withMessage('Completed must be a boolean'),
    body('priority')
      .optional()
      .isIn(['low', 'medium', 'high']).withMessage('Priority must be low, medium, or high')
  ],
  handleValidationErrors,
  async (req, res) => {
    try {
      const { title, description, completed, priority } = req.body;

      const updateData = {};
      if (title !== undefined) updateData.title = title;
      if (description !== undefined) updateData.description = description;
      if (completed !== undefined) updateData.completed = completed;
      if (priority !== undefined) updateData.priority = priority;

      const todo = await Todo.findByIdAndUpdate(
        req.params.id,
        updateData,
        { new: true, runValidators: true }
      );

      if (!todo) {
        return res.status(404).json({
          success: false,
          error: 'Todo not found'
        });
      }

      res.json({
        success: true,
        data: todo
      });
    } catch (error) {
      console.error('Error updating todo:', error);
      res.status(400).json({
        success: false,
        error: 'Error updating todo',
        message: error.message
      });
    }
  }
);

router.delete(
  '/:id',
  [
    param('id').isMongoId().withMessage('Invalid todo ID')
  ],
  handleValidationErrors,
  async (req, res) => {
    try {
      const todo = await Todo.findByIdAndDelete(req.params.id);

      if (!todo) {
        return res.status(404).json({
          success: false,
          error: 'Todo not found'
        });
      }

      res.json({
        success: true,
        message: 'Todo deleted successfully'
      });
    } catch (error) {
      console.error('Error deleting todo:', error);
      res.status(500).json({
        success: false,
        error: 'Server error while deleting todo',
        message: error.message
      });
    }
  }
);

module.exports = router;

File: backend/routes/todos.py
from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List
from bson import ObjectId
from datetime import datetime

from models.todo import TodoCreate, TodoUpdate, TodoResponse
from config.database import get_database

router = APIRouter()


def todo_helper(todo) -> dict:
    return {
        "_id": str(todo["_id"]),
        "title": todo["title"],
        "description": todo.get("description", ""),
        "completed": todo.get("completed", False),
        "priority": todo.get("priority", "medium"),
        "createdAt": todo.get("createdAt", datetime.utcnow()),
        "updatedAt": todo.get("updatedAt", datetime.utcnow())
    }


@router.get("/", response_model=dict)
async def get_todos(
    completed: Optional[bool] = None,
    priority: Optional[str] = None,
    sortBy: str = Query(default="createdAt"),
    order: str = Query(default="desc")
):
    try:
        db = get_database()
        filter_query = {}
        
        if completed is not None:
            filter_query["completed"] = completed
        
        if priority:
            filter_query["priority"] = priority
        
        sort_order = -1 if order == "desc" else 1
        
        todos_cursor = db.todos.find(filter_query).sort(sortBy, sort_order)
        todos = await todos_cursor.to_list(length=None)
        
        todos_list = [todo_helper(todo) for todo in todos]
        
        return {
            "success": True,
            "count": len(todos_list),
            "data": todos_list
        }
    except Exception as error:
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Server error while fetching todos",
                "message": str(error)
            }
        )


@router.get("/{id}", response_model=dict)
async def get_todo(id: str):
    try:
        if not ObjectId.is_valid(id):
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Todo not found"
                }
            )
        
        db = get_database()
        todo = await db.todos.find_one({"_id": ObjectId(id)})
        
        if not todo:
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Todo not found"
                }
            )
        
        return {
            "success": True,
            "data": todo_helper(todo)
        }
    except HTTPException:
        raise
    except Exception as error:
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Server error while fetching todo",
                "message": str(error)
            }
        )


@router.post("/", response_model=dict, status_code=201)
async def create_todo(todo: TodoCreate):
    try:
        if not todo.title:
            raise HTTPException(
                status_code=400,
                detail={
                    "success": False,
                    "error": "Title is required"
                }
            )
        
        db = get_database()
        todo_dict = {
            "title": todo.title,
            "description": todo.description or "",
            "completed": False,
            "priority": todo.priority or "medium",
            "createdAt": datetime.utcnow(),
            "updatedAt": datetime.utcnow()
        }
        
        result = await db.todos.insert_one(todo_dict)
        created_todo = await db.todos.find_one({"_id": result.inserted_id})
        
        return {
            "success": True,
            "data": todo_helper(created_todo)
        }
    except HTTPException:
        raise
    except Exception as error:
        raise HTTPException(
            status_code=400,
            detail={
                "success": False,
                "error": "Error creating todo",
                "message": str(error)
            }
        )


@router.put("/{id}", response_model=dict)
async def update_todo(id: str, todo_update: TodoUpdate):
    try:
        if not ObjectId.is_valid(id):
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Todo not found"
                }
            )
        
        db = get_database()
        existing_todo = await db.todos.find_one({"_id": ObjectId(id)})
        
        if not existing_todo:
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Todo not found"
                }
            )
        
        update_data = {}
        if todo_update.title is not None:
            update_data["title"] = todo_update.title
        if todo_update.description is not None:
            update_data["description"] = todo_update.description
        if todo_update.completed is not None:
            update_data["completed"] = todo_update.completed
        if todo_update.priority is not None:
            update_data["priority"] = todo_update.priority
        
        update_data["updatedAt"] = datetime.utcnow()
        
        await db.todos.update_one(
            {"_id": ObjectId(id)},
            {"$set": update_data}
        )
        
        updated_todo = await db.todos.find_one({"_id": ObjectId(id)})
        
        return {
            "success": True,
            "data": todo_helper(updated_todo)
        }
    except HTTPException:
        raise
    except Exception as error:
        raise HTTPException(
            status_code=400,
            detail={
                "success": False,
                "error": "Error updating todo",
                "message": str(error)
            }
        )


@router.delete("/{id}", response_model=dict)
async def delete_todo(id: str):
    try:
        if not ObjectId.is_valid(id):
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Todo not found"
                }
            )
        
        db = get_database()
        existing_todo = await db.todos.find_one({"_id": ObjectId(id)})
        
        if not existing_todo:
            raise HTTPException(
                status_code=404,
                detail={
                    "success": False,
                    "error": "Todo not found"
                }
            )
        
        await db.todos.delete_one({"_id": ObjectId(id)})
        
        return {
            "success": True,
            "message": "Todo deleted successfully"
        }
    except HTTPException:
        raise
    except Exception as error:
        raise HTTPException(
            status_code=500,
            detail={
                "success": False,
                "error": "Server error while deleting todo",
                "message": str(error)
            }
        )

